apps/browser/settings/declarativeloginmodel.cpp

```diff
@@ -0,0 +1,155 @@
+
+DeclarativeLoginModel::DeclarativeLoginModel(QObject *parent) : QAbstractListModel(parent)
+{
+    SailfishOS::WebEngine *webEngine = SailfishOS::WebEngine::instance();
+    webEngine->addObserver(ALL_LOGINS);
+    connect(webEngine, &SailfishOS::WebEngine::recvObserve, this, &DeclarativeLoginModel::handleRecvObserve);
+}
+
+void DeclarativeLoginModel::classBegin()
+{
+}
+
+void DeclarativeLoginModel::componentComplete()
+{
+    requestLogins();
+}
+
+void DeclarativeLoginModel::modify(int index, const QString &username, const QString &password)
+{
+    LoginInfo oldLogin = m_loginList.at(index);
+
+    if ((oldLogin.username() == username) && (oldLogin.password() == password)) {
+        return;
+    }
+
+    LoginInfo newLogin = oldLogin;
+    newLogin.setUsername(username);
+    newLogin.setPassword(password);
+
+    QVariantMap data;
+    data.insert(QStringLiteral("action"), "modify");
+    data.insert(QStringLiteral("oldinfo"), oldLogin.toMap());
+    data.insert(QStringLiteral("newinfo"), newLogin.toMap());
+
+    SailfishOS::WebEngine::instance()->notifyObservers(LOGINS_ACTION, QVariant(data));
+    m_loginList.replace(index, newLogin);
+
+    emit dataChanged(QAbstractListModel::index(index), QAbstractListModel::index(index), QVector<int>());
+}
+
+QVariant DeclarativeLoginModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.row() < 0 || index.row() >= m_loginList.count())
+        return QVariant();
+
+    switch (role) {
+    case Hostname:
+        return m_loginList.at(index.row()).hostname();
+    case Username:
+        return m_loginList.at(index.row()).username();
+    case Password:
+        return m_loginList.at(index.row()).password();
+    default:
+        return QVariant();
+    }
+}
+
+int DeclarativeLoginModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent)
+    return m_loginList.count();
+}
+
+QHash<int, QByteArray> DeclarativeLoginModel::roleNames() const
+{
+    QHash<int, QByteArray> roles;
+    roles[Hostname] = "hostname";
+    roles[Username] = "username";
+    roles[Password] = "password";
+    return roles;
+}
+
+int DeclarativeLoginModel::count() const
+{
+    return rowCount();
+}
+
+void DeclarativeLoginModel::handleRecvObserve(const QString &message, const QVariant &data)
+{
+    if (message == ALL_LOGINS) {
+        setLoginList(qvariant_cast<QVariantList>(data));
+    }
+}
+
+void DeclarativeLoginModel::setLoginList(const QVariantList &data)
+{
+    beginResetModel();
+
+    m_loginList.clear();
+
+    for (const auto &iter : data) {
+        QVariantMap varMap = iter.toMap();
+        m_loginList.append(LoginInfo(varMap));
+    }
+    endResetModel();
+    emit countChanged();
+}
+
+void DeclarativeLoginModel::requestLogins()
+{
+    QVariantMap data;
+    data.insert(QStringLiteral("action"), "getall");
+    SailfishOS::WebEngine::instance()->notifyObservers(LOGINS_ACTION, QVariant(data));
+}
+
+void DeclarativeLoginModel::remove(int index)
```

Comment:
If the user open context menu two times (while 5 seconds) and call remove, the second index will invalid or to point on other element. It could be avoided by searching the LoginInfo in m_loginList.
So, I see the 'getIndex()' still needed in this model, but it shouldn't use for removing elemenets from list. What do you think?

Reply:
Yes, a good point, thanks for spotting this. Unfortunately the LoginInfo items aren't necessarily unique, so it's not possible to find the right one just by searching on those. I've added a `uid` role to fix this.